extra_data/reader.py

```diff
@@ -88,7 +88,17 @@ def __init__(self, filename, _cache_info=None):
             self.control_sources = _cache_info['control_sources']
             self.instrument_sources = _cache_info['instrument_sources']
         else:
-            tid_data = self.file['INDEX/trainId'][:]
+            if self.format_version != '0.5':
+                # the INDEX/flag dataset contains a value (0 or 1) for each
+                # trainid recorded in the file. In normal case it is set to 1.
+                # If the DAQ receives a certain traindid from a data source
+                # before receiving it from the timeserver, it is assumed that
+                # the value for this trainid is wrong and we exclude it from
+                # the data collectiom.
+                valid_trains = self.file['INDEX/flag'][:].nonzero()[0].tolist()
+            else:
+                valid_trains = slice(None)
+            tid_data = self.file['INDEX/trainId'][valid_trains]
```

Comment:
Unfortunately, I think it will need several more changes in this file. We assume that the `.train_ids` array is aligned with data read from the indexes, e.g.

https://github.com/European-XFEL/EXtra-data/blob/27733a42bb3de41690f8648cd07ebcefda5f3dfa/extra_data/reader.py#L202-L204

https://github.com/European-XFEL/EXtra-data/blob/27733a42bb3de41690f8648cd07ebcefda5f3dfa/extra_data/reader.py#L592

https://github.com/European-XFEL/EXtra-data/blob/27733a42bb3de41690f8648cd07ebcefda5f3dfa/extra_data/reader.py#L1128-L1132

That's not an exhaustive list.

I think we'll need an additional attribute to select which values should be taken from indexes.

(This always relied on there being no 0s in `INDEX/trainId` except at the end. `extra-data-validate` checks for this, and I'm not aware that this has ever gone wrong.)